students/larry_mburu/Lesson04/mailroom.py

```diff
@@ -0,0 +1,233 @@
+
+    #add donation to donor history, in donation DB list
+    if donor_name in donations_per_individual: 
+        donations_per_individual[donor_name].append(donation_amount)
+
+    #send a thank you email
+    send_thank_you_email(donor_name, donation_amount)
+
+def create_report():
+    """
+    function creates a report based on donor database list. 
+    it calculates the total, average and number of gifts a 
+    donar has contributed
+    """
+    
+    #boarder column headers and divider. 
+    donor_name_header = "Donor Name"
+    total_give_header = "Total Given" 
+    num_gifts_header  = "Num Gifts"
+    average_gift_header = "Average Gift"
+    padding=20
+    boarder = '-'
+
+    print(f"{donor_name_header:<{padding}} |  {total_give_header} | {num_gifts_header} | {average_gift_header}")
+    print(boarder * 62)
+
+    # collection for donor statistical summaries for easy sorting
+    summary_of_donations = []
+
+    for donor in donations_per_individual:
+        donor_name = donor
+        total_given = sum(donations_per_individual[donor])
+        num_gifts = len(donations_per_individual[donor])
+        average_gift = total_given / num_gifts
+
+        summary_of_donations.append((donor_name, total_given, num_gifts, average_gift))
+
+    # sort summary_of_donations by average, which is indexed at 3 i.e 
+    # [("bill", 10, 2, 20.5)], save results in descending order.
+    summary_of_donations_sorted = sorted(summary_of_donations, key=itemgetter(3), reverse=True)
+
+    for summary in summary_of_donations_sorted: 
+        name = summary[0]
+        total = summary[1]
+        num_gifts = summary[2]
+        average = summary[3]
+        
+        print(f"{name:{padding}} $ {total:10.2f} {num_gifts:14} $ {average:10.2f}")
+
+    print()
+    # return to original prompt
+    main()
+def quit_program():
+    """
+    function quits the program, if the user selects
+    quit as an option in the menu display
+    """
+
+    print("Mail Automator (MA), signing off, Bye!")
+    sys.exit()
+
+def main(): 
+    """
+    main program entry prompt. prompts the user for an option in the menu
+
+    """
+    display_menu()
+
+    response = input("Choose one of the above options, [1..4]: ")
+    response = int(response)
+
+    # valid option numbers, per dipatch_dict
+    while response not in (1, 2, 3, 4): 
+        print("Invalid Option!")
+        response = input("Choose one of the following options, [1..4]: ")
+        response = int(response)
+
+    # dispatch dictionary. Key to function mapping. 
+    
+    dispatch_dict = { 
+        1 : prompt_for_donation_amount, 
+        2 : create_report, 
+        3 : send_thank_you_letter_to_all,
+        4 : quit_program
+    }
+
+    if response == 1: 
+        dispatch_dict.get(response)()
+    if response == 2: 
+        dispatch_dict.get(response)()
+    if response == 3: 
+        dispatch_dict.get(response)(validate=True)
+    if response == 4: 
+        dispatch_dict.get(response)()
```

Comment:
your above code needs a bit of refactor - it doesn't satisfy assignment requirements on how to implement switch dict but in general has some style issues.


here is an example
```
dispatch_dict = { 
        1 : prompt_for_donation_amount, 
        2 : create_report, 
        3 : send_thank_you_letter_to_all,
        4 : quit_program
    }

while True:
    response = input("...")
    if response not in dispatch_dict:
        print('invalid')
    dispatch_dict[response]()
```

let me know if you have any questions on the above code or why it's written this way.